scanpipe/pipes/d2d.py

```diff
@@ -387,3 +440,189 @@ def purldb_match(project, extensions, logger=None):
             start_time=start_time,
         )
         _resource_purldb_match(project, to_resource)
+
+
+def javascript_map(project, logger=None):
+    """Map a packed or minified JavaScript, TypeScript, CSS and SCSS to its source."""
+    project_files = project.codebaseresources.files().only("path")
+    query = Q(
+        *[
+            Q(path__endswith=extension)
+            for extension in PROSPECTIVE_JAVASCRIPT_MAP.keys()
+        ],
+        _connector=Q.OR,
+    )
+
+    from_resources = project_files.from_codebase().filter(query)
+    to_resources = project_files.to_codebase()
+    resource_count = from_resources.count()
+
+    if logger:
+        logger(
+            f"Mapping {resource_count:,d} from/ resources using javascript map "
+            f"against to/ codebase"
+        )
+
+    resource_iterator = from_resources.iterator(chunk_size=2000)
+    last_percent = 0
+    start_time = timer()
+    for resource_index, from_resource in enumerate(resource_iterator):
+        last_percent = pipes.log_progress(
+            logger,
+            resource_index,
+            resource_count,
+            last_percent,
+            increment_percent=10,
+            start_time=start_time,
+        )
+        _resource_javascript_map(from_resource, to_resources)
+
+
+def _resource_javascript_map(from_resource, to_resources):
```

Comment:
The code of this function is too complex with 3 imbricated `for` loops. 
It needs to be refactored into smaller pieces.